test/openssl/test_ssl.rb

```diff
@@ -1569,6 +1569,104 @@ def test_tmp_dh_callback
     end
   end
 
+  def test_ciphersuites_method_tls_connection
+    pend 'TLS 1.3 not supported' unless tls13_supported?
+    ssl_ctx = OpenSSL::SSL::SSLContext.new
+    tls13_ciphersuites = ssl_ctx.ciphers.select { |x| x[1] == 'TLSv1.3' }
+    if tls13_ciphersuites.empty? || !ssl_ctx.respond_to?(:ciphersuites=)
+      pend 'no support for TLSv1.3 cipher suites?'
+    end
+
+    tls13_csuite = tls13_ciphersuites[-1]
+    inputs = [tls13_csuite[0], [tls13_csuite[0]], [tls13_csuite]]
+
+    ctx_proc = -> ctx {
+      ctx.min_version = ctx.max_version = OpenSSL::SSL::TLS1_3_VERSION
+      ctx.ciphersuites = tls13_csuite[0]
+    }
+    start_server(ctx_proc: ctx_proc) do |port|
+      inputs.each do |input|
+        cli_ctx = OpenSSL::SSL::SSLContext.new
+        cli_ctx.min_version = cli_ctx.max_version = OpenSSL::SSL::TLS1_3_VERSION
+        cli_ctx.ciphersuites = input
+
+        server_connect(port, cli_ctx) do |ssl|
+          assert_equal(ssl.ssl_version, 'TLSv1.3')
+          assert_equal(ssl.cipher, tls13_csuite)
+        end
+      end
+    end
+  end
+
+  def test_ciphersuites_method_nil_argument
+    ssl_ctx = OpenSSL::SSL::SSLContext.new
+    pend 'ciphersuites= method is missing' unless ssl_ctx.respond_to?(:ciphersuites=)
+
+    assert_equal(ssl_ctx.ciphersuites = nil, nil)
+  end
+
+  def test_ciphersuites_method_frozen_object
+    ssl_ctx = OpenSSL::SSL::SSLContext.new
+    pend 'ciphersuites= method is missing' unless ssl_ctx.respond_to?(:ciphersuites=)
+
+    ssl_ctx.freeze
+    assert_raise(FrozenError) { ssl_ctx.ciphersuites = 'TLS_AES_256_GCM_SHA384' }
+  end
+
+  def test_ciphersuites_method_bogus_csuite
+    ssl_ctx = OpenSSL::SSL::SSLContext.new
+    pend 'ciphersuites= method is missing' unless ssl_ctx.respond_to?(:ciphersuites=)
+
+    assert_raise_with_message(
+      OpenSSL::SSL::SSLError,
+      /SSL_CTX_set_ciphersuites: no cipher match/i
+    ) { ssl_ctx.ciphersuites = 'BOGUS' }
+  end
+
+  def test_ciphers_method_tls_connection
+    ssl_ctx = OpenSSL::SSL::SSLContext.new
+    csuite = ssl_ctx.ciphers.select { |x| x[0] =~ /^(EC)?DHE?-RSA/ && \
+      x[1] == 'TLSv1.0' }[-1]
+
+    pend "couldn't find a TLSv1.0 cipher suite for testing" if csuite.nil?
```

Comment:
This seems too complex. Can this also use a hard-coded cipher suite, let's say, `ECDHE-RSA-AES256-GCM-SHA384`?

`start_server` uses RSA certificate, and we assume that OpenSSL supports ECDHE, so this cipher suite should work.

Reply:
How about ``['ECDHE-RSA-AES256-SHA384', 'TLSv1.2', 256, 256]``? The cipher suite you mention is from TLS1.3, I believe.

Reply:
I stand corrected -- it's a TLSv1.2 cipher suite. I still think my choice is fine. I should've known because of lists the key exchange and authentication methods!

Reply:
Running into an issue with openssl 1.0.2u test and some others. I'll try your suggestion. Also, the windows tests are failing on ``test_ciphersuites_method_tls_connection`` with an obscure error.

Reply:
Example of failure from Windows test:

```
1) Failure: test_ciphersuites_method_tls_connection(OpenSSL::TestSSL):
  exceptions on 1 threads:
  #<Thread:0x000001aa9d2ff910 D:/a/openssl/openssl/test/openssl/utils.rb:263 dead>:
  D:/a/openssl/openssl/lib/openssl/ssl.rb:524:in `accept': An established connection was aborted by the software in your host machine. - SSL_accept (Errno::ECONNABORTED)
  	from D:/a/openssl/openssl/lib/openssl/ssl.rb:524:in `accept'
  	from D:/a/openssl/openssl/test/openssl/utils.rb:238:in `block (3 levels) in start_server'
  	from D:/a/openssl/openssl/test/openssl/utils.rb:234:in `loop'
  	from D:/a/openssl/openssl/test/openssl/utils.rb:234:in `block (2 levels) in start_server'
D:/a/openssl/openssl/test/openssl/envutil.rb:281:in `assert_join_threads'
D:/a/openssl/openssl/test/openssl/utils.rb:289:in `block in start_server'
D:/a/openssl/openssl/test/openssl/utils.rb:210:in `pipe'
D:/a/openssl/openssl/test/openssl/utils.rb:210:in `start_server'
D:/a/openssl/openssl/test/openssl/test_ssl.rb:1581:in `test_ciphersuites_method_tls_connection'
     1578:     csuite = ['TLS_AES_128_GCM_SHA256', 'TLSv1.3', 128, 128]
     1579:     inputs = [csuite[0], [csuite[0]], [csuite]]
     1580: 
  => 1581:     start_server do |port|
     1582:       inputs.each do |input|
     1583:         cli_ctx = OpenSSL::SSL::SSLContext.new
     1584:         cli_ctx.min_version = cli_ctx.max_version = OpenSSL::SSL::TLS1_3_VERSION

Finished in 26.3875472 seconds.
```